pkg/autoscaler/aggregation/max/window.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package max
+
+type entry struct {
+	value float64
+	index int
+}
+
+// Window is a circular buffer which keeps track of the maximum value observed in a particular time.
+// Based on the "ascending minima algorithm" (http://web.archive.org/web/20120805114719/http://home.tiac.net/~cri/2001/slidingmin.html).
+type Window struct {
+	maxima        []entry
+	first, length int
+}
+
+// NewWindow creates an descending minima window buffer of size size.
+func NewWindow(size int) *Window {
+	return &Window{
+		maxima: make([]entry, size),
+	}
+}
+
+// Record records a value for a monotonically increasing index.
+func (m *Window) Record(index int, v float64) {
+	// Step One: Remove any elements where v > element.
+	// An element that's lower than the new element can never influence the
+	// maximum again, because the new element is both larger _and_ more
+	// recent than it.
+	originalLength := m.length
+	for i := 0; i < originalLength; i++ {
+		// Search backwards because that way we can delete by just decrementing length.
+		// The elements are guaranteed to be in descending order as described in Step Three.
+		if v >= m.maxima[m.index(m.first+originalLength-i-1)].value {
+			m.length--
+		} else {
+			// The elements are sorted, no point continuing.
+			break
+		}
+	}
+
+	// Step Two: Remove out of date elements from front of array.
+	// We only ever add at end of list, so the indexes are in ascending order,
+	// therefore the oldest are always first.
+	for m.length > 0 && index-m.maxima[m.first].index >= len(m.maxima) {
```

Comment:
Can we store the input `size` and use it to replace `len(m.maxima)` in this file?

Reply:
Theoretically, but then we get two fields in the struct called length (or one called length, one called size, which isn't much better!), which gets confusing. Empirically there's no appreciable efficiency difference here either way:

~~~~
benchmark                                                           old ns/op     new ns/op     delta
BenchmarkLargeTimeWindowRecord-16                                   69.9          71.6          +2.43%
BenchmarkLargeTimeWindowAscendingRecord-16                          38.9          38.7          -0.51%
BenchmarkLargeTimeWindowDescendingRecord/duration-5-minutes-16      47.9          49.0          +2.30%
BenchmarkLargeTimeWindowDescendingRecord/duration-15-minutes-16     49.1          48.8          -0.61%
BenchmarkLargeTimeWindowDescendingRecord/duration-30-minutes-16     49.0          48.3          -1.43%
BenchmarkLargeTimeWindowDescendingRecord/duration-45-minutes-16     49.0          48.5          -1.02%
~~~~